pkg/controllers/gateway/gateway_controller.go

```diff
@@ -97,12 +100,34 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		}
 		return ctrl.Result{}, nil
 	}
+	if !controllerutil.ContainsFinalizer(previous, GatewayFinalizer) && previous.GetDeletionTimestamp().IsZero() {
+		controllerutil.AddFinalizer(previous, GatewayFinalizer)
+		err = r.Update(ctx, previous)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
```

Comment:
This should probably all be moved to after the deletion steps https://github.com/Kuadrant/multicluster-gateway-controller/pull/162/files#diff-6afa7d73841a5c0ebdb78eaf859ce5a400203a037f871440940b5347c3a97376R111. I would have considered the first thing any reconciler should do is check if we are deleting it first.